### PR TITLE
Rating is coming to KooCook!

### PR DESCRIPTION
**Pull request by th-bunratta**
Today it's about time to introduce our new feature... star rating!
It introduces its widget, too.

New features are as follows:
- Users can now rate! (`91%` coverage)
  - Recipes
  - Posts
  - Comments

- Users can now follow each other, partially!

Apart from the enhancement feature, this introduces test coverage service provided by Codecov and more unit tests for Posts and Rating itself!

PR Prerequisite to these last three features:
- `feature/tag-label`
- `feature/popularity`
- `feature/personalisation`

or optionally,

- `feature/feedback`
- `feature/serving-size`

- `feature/ingredient-substitution`

### CODE REVIEW

File: koocook_core/models/review.py [hunk 1]
```diff
@@ -1,20 +1,34 @@
...
 
 
-def create_empty_aggregate_rating(**kwargs) -> 'AggregateRating':
-    """Creates an empty aggregate rating"""
-    return AggregateRating.objects.create(rating_value=0, rating_count=0, **kwargs)
+class ReviewableModel:
+    """
+        This creates an AggregateRating only after reviewables are actually created
+        and saved to a database; that is, it will be created once in need, not always.
+    """
```
Comment (MaiMee1) [inline koocook_core/models/review.py]: The docstring has one too many indent

File: koocook_core/models/review.py [hunk 2]
```diff
@@ -47,6 +60,10 @@ class Comment(SerialisableModel, models.Model):
         blank=True,
     )
 
+    def create_empty_aggregate_rating(**kwargs) -> 'AggregateRating':
+        """Creates an empty aggregate rating"""
+        return AggregateRating.objects.create(rating_value=0, rating_count=0, **kwargs)
+
```
Comment (MaiMee1) [inline koocook_core/models/review.py]: I don't think this is needed here?

File: koocook_core/models/review.py [hunk 3]
```diff
@@ -185,7 +197,10 @@ def add_rating(self, rating: Rating):
         self.check_rating(rating)
         total_value = self.rating_value * self.rating_count
         total_value += rating.rating_value
-        self.rating_count += 1
+        if not update:
+            self.rating_count += 1
+        else:
+            total_value -= rating.old_rating_value
         self.rating_value = total_value / self.rating_count
         self.save()
```
Comment (MaiMee1) [inline koocook_core/models/review.py]: Could you explain to me why it is done this way. This looks a but gimmicky.

Comment (th-bunratta) [reply koocook_core/models/review.py]: This is just like subtracting a sample from the summation of the average.

Comment (MaiMee1) [reply koocook_core/models/review.py]: ooh I get it now. There's a new 'update' function right. Kk

Comment (th-bunratta) [reply koocook_core/models/review.py]: It's an inverse approach to the `add_recipe` function.

**Comment by th-bunratta** [comment]
> Why are there so many merge requirements? Why aren't you branching off from `dev`? I don't think this is a sustainable workflow, especially if you aren't opening PRs right away.

You already repeated my reason why I did in this way; requirements are for branching from the main dev branch. Otherwise, I'd have to merge across feature branches.

**Comment by th-bunratta** [comment]
> Some issues:
> 
> * Stars light up when rating yourself and gives false cues even though there is a message showing that the rating is not successful
> * Rating doesn't transfer (or aren't implemented in Post Stream's "Trending")

Trending section is implemented in the successor branch `feature/personalisation`.

File: koocook_core/models/review.py (hunk 2 above)
Comment (th-bunratta) [reply koocook_core/models/review.py]: Totally

Comment (MaiMee1) [reply koocook_core/models/review.py]: Is using `AggregateRating.create_empty()` and still having this function the result of using changes' in `test/models`?

File: koocook_core/models/review.py (hunk 1 above)
Comment (th-bunratta) [reply koocook_core/models/review.py]: It's a visual indent. But for the convention's sake, I'll change it then.

File: koocook_core/models/review.py (hunk 2 above)
Comment (th-bunratta) [reply koocook_core/models/review.py]: I guess so.

Comment (MaiMee1) [reply koocook_core/models/review.py]: Just remove it then. It's going into dev so ut should be right. I don't care it's supposed to be changed anymore.

**Pull request merged**